### PR TITLE
[W3A-3] check if action source or dependencies have changes and skip upload

### PR DESCRIPTION
**Pull request by sl0b0dan**
check during action validation if source or dependencies have changes and skip uploading zips if they don't

### CODE REVIEW

File: rest/payloads/generated/actions/structs.conjure.go [hunk 1]
```diff
@@ -1371,8 +1365,20 @@ type TransactionPayload struct {
...
+	// Optional because it is added later so some payloads don't have it.
+	GasUsed *string `json:"gasUsed" conjure-docs:"Optional because it is added later so some payloads don't have it."`
+	// Optional because it is added later so some payloads don't have it.
+	CumulativeGasUsed *string `json:"cumulativeGasUsed" conjure-docs:"Optional because it is added later so some payloads don't have it."`
+	// Optional because it is added later so some payloads don't have it.
+	GasPrice *string `json:"gasPrice" conjure-docs:"Optional because it is added later so some payloads don't have it."`
+	// Optional because it is added later so some payloads don't have it.
+	GasTipCap *string `json:"gasTipCap" conjure-docs:"Optional because it is added later so some payloads don't have it."`
+	// Optional because it is added later so some payloads don't have it.
+	GasFeeCap *string `json:"gasFeeCap" conjure-docs:"Optional because it is added later so some payloads don't have it."`
```
Comment (BogdanHabic) [inline rest/payloads/generated/actions/structs.conjure.go:1380]: Should this be in the same PR?

File: commands/util/zip.go [hunk 2]
```diff
@@ -37,3 +39,23 @@ func MustZipDir(dirPath string, insidePath string, limitBytes int) []byte {
 
 	return content
 }
+
+func MustZipAndHashDir(dirPath string, insidePath string, limitBytes int) (*[]byte, *string) {
```
Comment (BogdanHabic) [inline commands/util/zip.go:43]: Maybe return values and point to them later?

File: commands/actions/publish.go [hunk 3]
```diff
@@ -317,12 +338,23 @@ func mustGetSource(sourcesDir string, locator string) string {
...
-		Actions: actions.ToRequest(sources),
+		Actions:             actions.ToRequest(sources),
+		LogicVersion:        nil,
+		DependenciesVersion: nil,
 	}
 
+	_, request.LogicVersion = util.MustZipAndHashDir(outDir, srcPathInZip, zipLimitBytes)
+
+	dependenciesDir := filepath.Join(actions.Sources, typescript.NodeModulesDir)
+	_, request.DependenciesVersion = util.ZipAndHashDir(dependenciesDir, nodeModulesPathInZip, zipLimitBytes)
```
Comment (BogdanHabic) [inline commands/actions/publish.go:356]: Why in one place we're using must and in the other just ZipAndHash?

Comment (BogdanHabic) [reply commands/actions/publish.go:356]: Also, won't there already be a dir check in `MustZipDir`?

File: commands/actions/publish.go [hunk 4]
```diff
@@ -254,15 +267,22 @@ func mustBuildProject(sourcesDir string, tsconfig *typescript.TsConfig) {
...
 	}
 }
 
-func mustValidateAndGetSources(r *rest.Rest, actions *actionsModel.ProjectActions, projectSlug string, sourcesDir string) map[string]string {
+func mustValidateAndGetSources(
+	r *rest.Rest,
+	actions *actionsModel.ProjectActions,
+	projectSlug string,
+	sourcesDir string,
+) (sources map[string]string, logicExist bool, dependenciesExist bool) {
```
Comment (BogdanHabic) [inline commands/actions/publish.go:285]: Removed named parameters in favor of docs.

**Review by BogdanHabic — COMMENTED** [review]
Added comments.

File: rest/payloads/generated/actions/structs.conjure.go (hunk 1 above)
Comment (sl0b0dan) [reply rest/payloads/generated/actions/structs.conjure.go:1380]: Those fields are generated from conjure earlier, but not copied to tenderly-cli at that moment as they are not used here, so they ended up in this PR with other changes in the generated model. Do you think I should remove them from this PR?

File: commands/actions/publish.go (hunk 3 above)
Comment (sl0b0dan) [reply commands/actions/publish.go:356]: `MustZipAndHashDir` is used for the action source code which must exist and `ZipAndHashDir` is used for the node_modules folder, which doesn't need to be present. 
`MustZipDir` is checking if dir exists and exiting if it doesn't, which is not suitable for node_modules.
Do you have any hint for refactoring these so they are usable in both cases?

**Review by asalem1987 — APPROVED** [review]

**Review by superstrar23 — APPROVED** [review]

**Comment by sl0b0dan** [comment]
this should wait for changes from [tenderly-server PR](https://github.com/Tenderly/tenderly-core/pull/569) to be merged and deployed

File: commands/actions/publish.go [hunk 5]
```diff
@@ -272,7 +287,7 @@ func mustValidateAndGetSources(r *rest.Rest, actions *actionsModel.ProjectAction
 		validatedSources[name] = source
 	}
 
-	mustValidate(r, actions, validatedSources, projectSlug)
+	logicExist, dependenciesExist = mustValidate(r, actions, validatedSources, projectSlug)
```
Comment (veljko-matic) [inline commands/actions/publish.go:290]: I'll maybe refactor this not to use global vars...

**Review by veljko-matic — COMMENTED** [review]
Need more context of this task, but on brief look LGTM

**Review by mspuz — APPROVED** [review]

**Pull request merged**